### ISSUE TITLE
Using CSV to store glyph metadata

### ISSUE THREAD

**Issue by madig** [googlefonts/googlesans#450]
Inspired by https://github.com/unified-font-object/ufo-spec/issues/163 I wrote https://gist.github.com/madig/94bbb509e8458654bfa941f0368b53d9 to dump UFO glyph data into a CSV file. Excel likes to interpret even quoted things as numbers, so I used https://winmerge.org/ to look at the data dump for uprights and italics side-by-side. Found various mismatches I'm not sure about.

[glyph_data_upright.csv](https://github.com/googlefonts/googlesans/files/7932684/glyph_data_upright.csv)
[glyph_data_italics.csv](https://github.com/googlefonts/googlesans/files/7932685/glyph_data_italics.csv)

Interesting though. At the time of this writing, the script won't dump the CSV data back into UFOs, but I can see it helping us.

One could also split the data up by script and merge during the build script to populate the UFOs before passing them to the compiler. That'd help with data management.

Edit: btw, https://github.com/BurntSushi/xsv seems like a nice CLI tool to query, slice, tabulate and otherwise handle CSV files.

Edit 2: Watch this (fish shell, comparing data for uprights and italics):

```
> diff -U2 (xsv table /tmp/glyph_data_uprights.csv | psub) (xsv table /tmp/glyph_data_italics.csv | psub)
--- /tmp/.psub.oqpsoyMwzL       2022-01-25 11:17:45.420489005 +0000
+++ /tmp/.psub.dKlqPSe9Ob       2022-01-25 11:17:45.420489005 +0000
@@ -87,7 +87,7 @@
 Char-georgian                                  uni1CAD                        1CAD                         True
 Che-cy                                         uni0427                        0427      base               True
-Cheabkhasian-cy                                uni04BC                        04BC                         True
+Cheabkhasian-cy                                uni04BC                        04BC      base               True
 Chedescender-cy                                uni04B6                        04B6      base               True
-Chedescenderabkhasian-cy                       uni04BE                        04BE                         True
+Chedescenderabkhasian-cy                       uni04BE                        04BE      base               True
 Chedieresis-cy                                 uni04F4                        04F4      base               True
 Cheh-arm                                       uni0543                        0543                         True
@@ -321,5 +321,5 @@
 Itilde                                                                        0128      base               True
 Itilde.alt                                                                              base               False
-Iu-cy                                          uni042E                        042E      base               True
+Iu-cy                                          uni042E                        042E                         True
 Izhitsa-cy                                     uni0474                        0474                         True
 J                                                                             004A      base               True
[...]
```

Neat!

**Comment by chrissimpkins** [comment]
Very nice!  I really like that tabular diff with xsv output.

> One could also split the data up by script and merge during the build script to populate the UFOs before passing them to the compiler. That'd help with data management.

Does this mean that "production UFO sources" as committed to the repository would not (necessarily) have valid data?  Valid data are populated in temp UFO source files from the CSV definitions at build time?  If so, I think that this means that we would need an approach to export valid data in UFO spec files that designers can work with in their editors.

This looks good for analytics / QA across the glyph set in addition to any build pipeline improvements that it might offer!

**Comment by madig** [comment]
> Does this mean that "production UFO sources" as committed to the repository would not (necessarily) have valid data? Valid data are populated in temp UFO source files from the CSV definitions at build time?

I'd rather have a separate format that can export valid standalone UFOs.

**Comment by chrissimpkins** [comment]
> > Does this mean that "production UFO sources" as committed to the repository would not (necessarily) have valid data? Valid data are populated in temp UFO source files from the CSV definitions at build time?
> 
> 
> 
> I'd rather have a separate format that can export valid standalone UFOs.

Interesting, so "prod sources" as maintained in repo would be something, as yet, unspec'd.  Look forward to seeing what you come up with!

**Comment by madig** [comment]
Yes. I don't know yet how it looks and if it would work at all. I'll continue to mull over it.

If it fails, we could maintain sets of UFOs and combine them via script before running fontmake on them. Not as clean, but probably doable.

**Comment by belluzj** [comment]
Is this still relevant for GS, now that the general direction seems to be to use glyphspackage in GS projects?
